### ISSUE TITLE
Sorting of users

### ISSUE THREAD

**Issue by KB7777** [nextcloud/server#3230]
It would be great to have sorting users feature by the last login date or quota set. LDAP users either :)

Regards.

**Comment by MorrisJobke** [comment]
cc @blizzz @jancborchardt

**Comment by MorrisJobke** [comment]
ref #117

**Comment by jancborchardt** [comment]
Yup, should be possible to sort by the headers where it makes sense. For those two columns for sure. :)

@blizzz @icewind1991

**Comment by blizzz** [comment]
**TL;DR:** The only valid approach to this is to Client-side sorting in Javascript, with the disadvantage that you do not have complete results, only that what you already see. If this is good enough, should not be much of a problem.

**Long story**

How its describes is also how sorting in Files work. With the difference that all files are retrieved in background. Which is not possible for users (too many, far too slow).

In general, you can pick between two approaches:

**Client side sorting**

Advantage: fast, no server interaction

Disadvantage: may be incomplete, sorts only the rows that are present

**Server side sorting**

Background: for the provided cols, looking at preferences table is sufficient (needs to ask user backends for meta data still). 

Advantage: If trying hard enough, you get a complete picture…

Disadvantage: requires server interaction → slow. Won't retrieve users without rows set, (unless you ask user backends first for all! users – no-go on big instances). Even slower (must match results from preferences table and user backends), and most likely breaks paging, is the combination with a search filter (you will have gaps and never know whether there are more users to retrieve).

**Comment by alexanderdd** [comment]
From what I read, I think it should be **server side**, so it also works for big instances (including LDAP). 

It's better to have a feature that works for all instances, including big ones, instead of having a feature (client side sorting) that only works well on small instances where only the rows that you already see are sorted.

Background: Sorting by last login is an important feature for admins, especially looking at those that never logged in, cause you might wanna delete them/contact them.

**Comment by alexanderdd** [comment]
**Please** implement this feature.

I just wanted to remove old users from my intance and can still not sort by last login date, Nextcloud 23.0.8

@skjnldsv @MorrisJobke ? Or who should I tag?

**Comment by jospoortvliet** [comment]
Hi @alexanderdd, 
pinging people is sadly not really what gets it implemented ;-)

Somebody doing it in their free time is, or somebody getting paid to do it. This doesn't sound to me like a feature a volunteer would work on as it's not really a home user thing, so your best bet is probably to either pay somebody or do it yourself.

Either way, please avoid bothering people about it - that's just not polite and certainly doesn't create a motivation to help. There are thousands of other things that need doing I'm afraid and most of those have far more than 8 👍🏻  and user impact.

I hope this'll work out for you.

**Comment by joshtrichards** [comment]
Related PR: #45245

**Comment by Pytal** [comment]
A separate "Recently active" section has been added which sorts users by last login: https://github.com/nextcloud/server/pull/45716

If sorting by a different field is desired feel free to open a new issue :)

**Comment by alexanderdd** [comment]
@Pytal thank you for working on this! In a 29.0.3 test install, I see the Last login column, but there is no way to sort ascending/descending... did I miss something? Or will it come in NC 30?

(I only have two users on that instance, does sorting only work when there are many users? On another 28.0.7 install with many users, there is no sorting)

**Comment by Pytal** [comment]
> @Pytal thank you for working on this! In a 29.0.3 test install, I see the Last login column, but there is no way to sort ascending/descending... did I miss something? Or will it come in NC 30?

You're welcome :)

As said this is only for NC30 and is actually a separate section see:
![image](https://github.com/user-attachments/assets/05bcbe70-4fc3-417e-a224-63ec63a46cea)

**Comment by alexanderdd** [comment]
But then it does not close this issue here, because it asks for "sorting users feature by the last login date or quota set", I think what is asked (and what I also find useful) is that in the regular "all accounts" section, you click on one of the headers of a table column, and it sorts the users by that column, e.g. last login date, quota, language, ... It should be server side sorting (see discussion above).

I think I will reopen this issue, unless you protest?!

**Comment by Pytal** [comment]
Due to technical limitations with e.g. LDAP it was concluded that having last login sort in a separate section rather than the header buttons is the best and cleanest way

For other fields it may be feasible. Feel free to open a separate issue for those!

**Comment by alexanderdd** [comment]
I will reopened this issue because this is about "sorting of users", for example "by the last login date or quota set" (first post).

NC 30 introduced a new sidebar "recently active", which is no sorting of users but a custom filter (it is also not clear where the cutoff is? Is 5 years ago still "recently active"? why/why not?) In particular, I would be interested in the users that never logged in, or logged in more than one/two/five/.. year ago, and this new sidebar doesn't help with that.

The PR also says https://github.com/nextcloud/server/pull/45249 "sorting of users" but it is not actually sorting, as explained.

The [release notes](https://help.nextcloud.com/t/updated-summary-of-what-s-new-in-nextcloud-version-30-human-readable/205018) call this "Sort Users by Last Login: Administrators can sort user lists by last login time to manage inactive accounts effectively." which is not what the new feature does.

I still believe from a UX perspective, the best would be to allow for ascending/descending sorting by any attribute (title of the column). If it takes a long time, just warn the user "sorting can take up to 15mins - continue?". Or if there is a decision not to implement this, this should be "close as not planned".

**Comment by S4enL0ng** [comment]
Is it at least possible to make this sorting on the occ commands? for example only if used with "--limit 0" like in `occ user:lastseen --all --limit 0 --sort 'date'` or something.
